File: framework/languages/node/e2e/PubSub/Client/Support/subscriber-observation.ts
import { getJson, postJsonWithin } from '../../../http-client';
import { PubSubNames } from '../../Shared/messages';
import { delay } from './scenario-assert';

export async function waitForEvent(
  subscriberUrl: string,
  runId: string,
  sequence: number,
  value: string,
  timeoutMilliseconds = 10_000
): Promise<readonly string[]> {
  return await postJsonWithin<readonly string[]>(subscriberUrl, '/evidence/wait', {
    containsAllLineGroups: [[
      'event|',
      `run=${runId}`,
      `topic=${PubSubNames.mainTopic}`,
      `seq=${sequence}|`,
      `value=${value}`
    ]],
    timeoutMilliseconds
  }, timeoutMilliseconds + 1_000);
}

export async function waitForNoEvent(
  subscriberUrl: string,
  runId: string,
  sequence: number,
  value: string,
  timeoutMilliseconds = 1_000
): Promise<void> {
  const deadline = Date.now() + timeoutMilliseconds;
  while (Date.now() < deadline) {
    const evidence = await getJson<readonly string[]>(subscriberUrl, '/evidence');
    if (evidence.some((line) => line.includes('event|')
      && line.includes(`run=${runId}`)
      && line.includes(`seq=${sequence}|`)
      && line.includes(`value=${value}`))) {
      throw new Error(`Subscriber received event while it was expected to be disconnected: run=${runId}.`);
    }
    await delay(100);
  }
}

File: framework/languages/dotnet/e2e/PubSub/Client/Support/SubscriberObservation.cs
using PubSub.Shared;
using Zlink.HttpClient;

namespace PubSub.Client.Support;

internal static class SubscriberObservation
{
    public static async Task<int> EvidenceCountAsync(ZLinkHttpClient subscriber)
    {
        return (await subscriber.Get("/evidence").Async<string[]>()).Body.Length;
    }

    public static async Task<string[]> WaitForEventAsync(
        ZLinkHttpClient subscriber,
        string runId,
        int sequence)
    {
        return (await subscriber.Post("/evidence/wait")
            .Body(new EvidenceWaitReq([], [])
            {
                ContainsAllLineGroups =
                [
                    [
                        "event|",
                        $"run={runId}",
                        $"topic={PubSubNames.MainTopic}",
                        $"seq={sequence}|"
                    ]
                ]
            })
            .Async<string[]>()).Body;
    }
}
